File: Leetcode/03_Hash-Table/022_0018_4sum.go
// https://leetcode.cn/problems/4sum/
// m

package main

import "sort"

// same as 3sum.
func fourSum(nums []int, target int) (res [][]int) {
	sort.Ints(nums)
	for i := 0; i < len(nums)-3; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		for j := i + 1; j < len(nums)-2; j++ {
			if j > i+1 && nums[j] == nums[j-1] {
				continue
			}
			l, r := j+1, len(nums)-1
			jackpot := target - nums[i] - nums[j]
			for l < r {
				sum := nums[l] + nums[r]
				if sum < jackpot {
					l++
				} else if sum > jackpot {
					r--
				} else {
					res = append(res, []int{nums[i], nums[j], nums[l], nums[r]})
					l++
					r--
					for l < r && nums[l] == nums[l-1] {
						l++
					}
					for l < r && nums[r] == nums[r+1] {
						r--
					}
				}
			}
		}
	}
	return
}

File: Leetcode/03_Hash-Table/022_0018_4sum.py
# https://leetcode.cn/problems/4sum/
# m

class Solution(object):
    def fourSum(self, nums, target):
        nums.sort()
        res = []
        for i in range(len(nums)-3):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            for j in range(i+1, len(nums)-2):
                if j > i+1 and nums[j] == nums[j-1]:
                    continue
                jackpot = target - nums[i] - nums[j]
                l, r = j+1, len(nums)-1
                while l < r:
                    _sum = nums[l] + nums[r]
                    if _sum > jackpot:
                        r -= 1
                    elif _sum < jackpot:
                        l += 1
                    else:
                        res.append([nums[i], nums[j], nums[l], nums[r]])
                        l += 1
                        r -= 1
                        while l < r and nums[l] == nums[l-1]:
                            l += 1
                        while l < r and nums[r] == nums[r+1]:
                            r -= 1
        return res